### PR TITLE
[WIP] Core Desktop Serialization Roundtrip work

### PR DESCRIPTION
**Pull request by krwq**
Stub for out serialization NetFX work - work in progress

Types with failures when deserializing netfx->core:
- [ ] System.Net.CookieCollection
- [ ] System.Net.CookieContainer
- [ ] System.Collections.BitArray
- [ ] System.Collections.ArrayList
- [ ] System.Collections.Comparer
- [ ] System.Collections.Queue
- [ ] System.Collections.Generic.SortedDictionary
- [ ] System.Collections.SortedList
- [ ] System.Collections.Hashtable
- [ ] System.ComponentModel.BindingList
- [ ] System.Collections.Stack
- [ ] System.Collections.Specialized.StringDictionary
- [ ] System.Collections.Specialized.StringCollection
- [ ] System.Collections.Specialized.HybridDictionary
- [ ] System.Collections.ObjectModel.ObservableCollection
- [ ] System.Collections.ObjectModel.ReadOnlyObservableCollection
- [ ] System.Collections.ObjectModel.ReadOnlyDictionary
- [ ] System.Runtime.Serialization.Formatters.Tests.SimpleKeyedCollection

cc: @ViktorHofer
[log.txt](https://github.com/dotnet/corefx/files/1046203/log.txt)

### CODE REVIEW

File: src/System.ObjectModel/src/System/Collections/ObjectModel/ReadOnlyObservableCollection.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@ namespace System.Collections.ObjectModel
     [Serializable]
     [DebuggerTypeProxy(typeof(CollectionDebugView<>))]
     [DebuggerDisplay("Count = {Count}")]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public class ReadOnlyObservableCollection<T> : ReadOnlyCollection<T>, INotifyCollectionChanged, INotifyPropertyChanged
```
Comment (justinvp) [inline src/System.ObjectModel/src/System/Collections/ObjectModel/ReadOnlyObservableCollection.cs]: In the [reference source](https://github.com/Microsoft/referencesource/blob/90b323fe52bec428fe4bd5f007e9ead6b265d553/System/compmod/system/collections/objectmodel/readonlyobservablecollection.cs#L28), this has `[TypeForwardedFrom("WindowsBase, Version=3.0.0.0, Culture=Neutral, PublicKeyToken=31bf3856ad364e35")]`. Based on @weshaggard's [comment](https://github.com/dotnet/coreclr/pull/12059#discussion_r119905431), I assume it should match here.

File: src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs [hunk 2]
```diff
@@ -17,6 +17,7 @@ namespace System.Collections.ObjectModel
     [Serializable]
     [DebuggerTypeProxy(typeof(CollectionDebugView<>))]
     [DebuggerDisplay("Count = {Count}")]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, PublicKeyToken=b77a5c561934e089")]
```
Comment (justinvp) [inline src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs]: Same comment as https://github.com/dotnet/corefx/pull/20599/files#r119907105

File: src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs [hunk 3]
```diff
@@ -349,6 +350,7 @@ private void OnCollectionReset()
 
         // this class helps prevent reentrant calls
         [Serializable]
+        [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, PublicKeyToken=b77a5c561934e089")]
```
Comment (justinvp) [inline src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs]: Same comment as https://github.com/dotnet/corefx/pull/20599/files#r119907105

**Comment by ViktorHofer** [comment]
@justinvp thanks for the great help! Have you checked all of them? We have tests which will fail if we have some mismatches but to catch them early we should make sure there aren't any other typeforwards in netfx.

**Comment by justinvp** [comment]
> Have you checked all of them?

@ViktorHofer, I didn't do an exhaustive review. I just did a quick scan and pointed out the ones I'm aware of.

**Comment by ViktorHofer** [comment]
I just checked 50% of them and they seem safe

**Comment by krwq** [comment]
@ViktorHofer that should fix your `PlatformDetection.IsNetfxBelow471()` issues

**Comment by safern** [comment]
> cc: @safern - please take a look at PlatformDetection logic

This is the "correct" solution to get the actual framework version but the problem is that RegistryKey apis are not available in uapaot so the uapaot tests wont build. As you can see the vertical build just failed in this PR. I actually opened a PR for this last week and encountered the same issue: https://github.com/dotnet/corefx/pull/20213

I already have another solution in progress which is mapping the actual build number to the correct version by reading the build number from `RuntimeInformation.FrameworkDescription`. 

Will submit a PR to unblock this.

**Comment by krwq** [comment]
ok, undid that commit, let's wait for @safern's PR

**Comment by danmosemsft** [comment]
@dotnet-bot help please

**Comment by ViktorHofer** [comment]
Closing in favor of https://github.com/dotnet/corefx/pull/20697

**Pull request closed without merging**